Week4/HW04.hs

```diff
@@ -0,0 +1,149 @@
+{-# OPTIONS_GHC -Wall #-}
+
+-- https://www.cis.upenn.edu/~cis194/spring15/hw/04-typeclasses.pdf
+
+module Week4.HW04 where
+
+import Data.List (dropWhileEnd)
+newtype Poly a = P [a]
+
+-- Exercise 1 -----------------------------------------
+
+x :: Num a => Poly a
+x = P [0, 1]
+
+-- Exercise 2 ----------------------------------------
+
+instance (Num a, Eq a)              => Eq (Poly a) where
+    -- This will not be effecient way of doing it I think,
+    -- because its taking extra storage to save the new list 
+    -- without trailing zeros and then does the matching
+    P a == P b                      = withoutTrailingZerosA == withoutTrailingZerosB
+        where 
+            withoutTrailingZerosA   = dropWhileEnd (== 0) a 
+            withoutTrailingZerosB   =  dropWhileEnd (== 0) b
+        
+
+-- Exercise 3 -----------------------------------------
+
+-- e == exponent 
+-- c == count of the polynomial term, is it first, second, third and so on
+showHelper :: (Num a, Show a, Eq a) => Poly a -> Int -> Int -> String
+showHelper (P (p:ps)) e c
+    | p == 0                = showHelper (P ps) (e-1) (c+1)
+    | e == 0                = leadingPlus ++ show p 
+    -- I was trying to get nested guards here instead of using [if-then-else] or [case] but wasnt able to 
+    | e == 1                = leadingPlus ++ leadingNum ++ "x"  ++ showHelper (P ps) (e-1) (c+1)
+    | otherwise             = leadingPlus ++ leadingNum ++ "x^" ++ show e ++ showHelper (P ps) (e-1) (c+1)
+    where 
+        leadingPlus         = case c of 
+                0           -> "" 
+                _           -> " + "
+        leadingNum          = case p of 
+                1           -> ""
+                anyP        -> show anyP
+
+
+showHelper (P []) _ _       = ""
+
+instance (Num a, Eq a, Show a) => Show (Poly a) where
+    show (P p) = 
+        showHelper (P $ reverse withoutTrailingZeros ) (length withoutTrailingZeros - 1) 0
+            where 
+                withoutTrailingZeros = dropWhileEnd (==0) p
+
+-- Exercise 4 -----------------------------------------
+
+plus :: Num a               => Poly a -> Poly a -> Poly a
+plus (P polyA) (P polyB)    = P $ zipWith (+) (polyA ++ replicate leadingZerosToA 0) (polyB ++ replicate leadingZerosToB 0)
```

Comment:
This a good example when defining custom recursion function might make the code a bit cleaner.
But your way is fine too (except it won't work with infinite list, for which `length` will end up in infinite loop)

```haskell
plus (P as) (P bs) = P (help as bs)
  where
    help as [] = as
    help [] bs = bs
    help (a:as) (b:bs) = a+b:help as bs
```